sycl/include/sycl/ext/oneapi/bfloat16.hpp

```diff
@@ -141,10 +141,21 @@ class bfloat16 {
 private:
   Bfloat16StorageT value;
 
+  // Private tag used to avoid constructor ambiguity.
+  struct private_tag {
+    explicit private_tag() = default;
+  };
+
+  constexpr bfloat16(Bfloat16StorageT Value, private_tag) : value{Value} {}
+
   // Explicit conversion functions
   static float to_float(const Bfloat16StorageT &a);
   static Bfloat16StorageT from_float(const float &a);
 
+  // Friend traits.
+  friend std::numeric_limits<bfloat16>;
+  friend std::hash<bfloat16>;
```

Comment:
Do we have any guarantees on the layout/size? If so, can we use `bit_cast` instead of friendship?

Reply:
We do, but `bit_cast` in our implementation is only _sometimes_ `constexpr`, so the friendship for `numeric_limits` lets it be unconditionally `constexpr`. For `std::hash` it should be fine to drop the friendship though!